pym/bob/cmds/build/builder.py

```diff
@@ -104,6 +108,100 @@ class RestartBuildException(Exception):
 class CancelBuildException(Exception):
     pass
 
+class JobServer:
+    def jobserver_thread(self):
+        while self.__running:
+            os.write(self.__w_server, os.read(self.__r_server, 1))
+
+    def __init__(self, jobs):
+        self.__running = True
+        self.thread = threading.Thread(target=self.jobserver_thread, args =())
+        self.__r_make, self.__w_server = os.pipe()
+        self.__r_server, self.__w_make = os.pipe()
+        os.write(self.__w_server, bytes(jobs))
+        self.thread.start()
+
+    def getMakeFd(self):
+        return [ self.__r_make, self.__w_make ]
+
+    def stop(self):
+        self.__running = False
+        os.write(self.__w_make, bytes(0x30))
+        self.thread.join()
+
+
+class JobServerSemaphore:
+    def __init__(self, value, fds):
+        if sys.platform == "win32":
+            self.__sem = asyncio.BoundedSemaphore(value)
+        else:
+            self.__sem = asyncio.Semaphore(0)
+            self.__lock = threading.Lock()
+            self.__waitersCnt = 0
+            self.__jobsAquired = 0
+            self.__fds = fds
+            fcntl.fcntl(self.__fds[0], fcntl.F_SETFL,
+                    fcntl.fcntl(self.__fds[0], fcntl.F_GETFL) | os.O_NONBLOCK)
+
+    @staticmethod
+    def jobavailableCallback(self):
+        ret = None
+        try:
+            ret = os.read(self.__fds[0], 1)
+        except BlockingIOError:
+            pass
+        if ret:
+            with self.__lock:
+                self.__waitersCnt -= 1
+                if self.__waitersCnt == 0:
+                    asyncio.get_event_loop().remove_reader(self.__fds[0])
+            self.__sem.release()
```

Comment:
What about this?

```suggestion
        try:
            while __waitersCnt:
                os.read(self.__fds[0], 1)
                self.__waitersCnt -= 1
                self.__sem.release()
        except BlockingIOError:
            pass
        if self.__waitersCnt == 0:
            asyncio.get_event_loop().remove_reader(self.__fds[0])
```

What is still not right is that we have to remember the values that were extracted from the pipe and use them when writing back to the pipe.

Reply:
Haven't tested this but I think it will not work as os.read is not blocking? So when ever a token is back in the pipe all waiters are released, aren't they? At least the return of `os.read` must be checked?

Reply:
My assumption was that either we get exactly one byte from the pipe (and hence we got a ticket) or `read` will throw `BlockingIOError` because the pipe was empty and the call would have blocked. If this assumption does not hold then the code above is indeed garbage.